mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,227 @@
+use crate::token_source::{SasTokenSource, TokenSource};
+
+const DEFAULT_TOKEN_DURATION_MINS: u64 = 60;
+const DEFAULT_TOKEN_DURATION_SECS: Duration = Duration::from_secs(60 * DEFAULT_TOKEN_DURATION_MINS);
+const DEFAULT_MAX_RECONNECT: Duration = Duration::from_secs(5);
+
+#[derive(Debug)]
+pub struct ShutdownHandle(mqtt3::ShutdownHandle);
+
+impl ShutdownHandle {
+    pub async fn shutdown(&mut self) -> Result<(), ClientConnectError> {
+        self.0
+            .shutdown()
+            .await
+            .map_err(ClientConnectError::ShutdownClient)?;
+        Ok(())
+    }
+}
+
+pub struct TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    address: String,
+    token_source: Option<T>,
+}
+
+impl<T> IoSource for TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    type Io = TcpStream;
+    type Error = std::io::Error;
+    type Future = BoxFuture<'static, Result<(TcpStream, Option<String>), std::io::Error>>;
+
+    fn connect(&mut self) -> Self::Future {
+        let address = self.address.clone();
+        let token_source = self.token_source.as_ref().cloned();
+
+        Box::pin(async move {
+            let expiry = Utc::now()
+                + chrono::Duration::from_std(DEFAULT_TOKEN_DURATION_SECS).unwrap_or_else(|_| {
+                    chrono::Duration::minutes(DEFAULT_TOKEN_DURATION_MINS.try_into().unwrap())
+                });
+            let mut password = None;
+            if let Some(ts) = token_source {
+                password = match ts.get(&expiry).await {
+                    Ok(x) => Some(x),
+                    Err(e) => {
+                        error!("Failed to get token for connection {} {}", address, e);
+                        None
+                    }
+                }
+            };
+
+            let io = TcpStream::connect(address).await;
+            io.map(|io| (io, password))
+        })
+    }
+}
+
+pub struct MqttClient<T>
+where
+    T: EventHandler,
+{
+    client: Client<TcpConnection<SasTokenSource>>,
+    event_handler: T,
+}
+
+impl<T: EventHandler> MqttClient<T> {
+    pub fn new(
+        address: &str,
+        keep_alive: Duration,
+        clean_session: bool,
+        event_handler: T,
+        connection_credentials: &Credentials,
+    ) -> Self {
+        let (client_id, token_source) = match connection_credentials {
+            Credentials::Provider(provider_settings) => (
+                provider_settings.device_id().into(),
+                Some(SasTokenSource::new(connection_credentials.clone())),
+            ),
+            Credentials::PlainText(creds) => (
+                creds.client_id().into(),
+                Some(SasTokenSource::new(connection_credentials.clone())),
+            ),
+            Credentials::Anonymous(client_id) => (client_id.into(), None),
+        };
+
+        let client = if clean_session {
+            Client::new(
+                Some(client_id),
+                None,
```

Comment:
I think it wants some `username`

Reply:
this is the same as for command handler, only client_id is specified